onnxruntime/core/providers/nuphar/scripts/model_editor.py

```diff
@@ -0,0 +1,700 @@
+            # only support 1 scan input
+            assert num_scan_inputs == 1
+            new_inputs = list(in_n.input)
+            in_sg_inputs = [i.name for i in in_sg.input]
+            replaced_matmul = None
+            for in_sn in in_sg.node:
+                if in_sn.op_type == 'Concat' and all([i in in_sg_inputs for i in in_sn.input]):
+                    # make sure the concat's inputs are scan input and scan state
+                    assert NodeFactory.get_attribute(in_sn, 'axis') == 1
+                    assert in_sn.input[0] == in_sg_inputs[-1]
+                    matmul_node = [nn for nn in in_sg.node if nn.op_type == 'MatMul' and in_sn.output[0] in nn.input]
+                    assert matmul_node
+                    replaced_matmul = matmul_node[0]
+                    assert replaced_matmul.input[1] in initializers
+                    aa = nf.get_initializer(replaced_matmul.input[1])
+                    input_size = in_sg.input[-1].type.tensor_type.shape.dim[-1].dim_value
+                    input_proj_weights, hidden_proj_weights = np.vsplit(aa, [input_size])
+                    # add matmul for input_proj outside of Scan
+                    input_proj = nf.make_node('MatMul', [new_inputs[-1], input_proj_weights])
+                    new_inputs[-1] = input_proj.name
+                    out_sg.input[-1].type.tensor_type.shape.dim[-1].dim_value = input_proj_weights.shape[-1]
+                    # add matmul for hidden_proj inside Scan
+                    hidden_proj = nf_subgraph.make_node('MatMul', [in_sn.input[1], hidden_proj_weights])
+                    nf_subgraph.make_node('Add', [out_sg.input[-1].name, hidden_proj], output_names=replaced_matmul.output[0])
+                    # remove initializer of concat matmul
+                    if not [n for n in in_mp.graph.node if n != in_n and replaced_matmul.input[1] in n.input]:
+                        nf.remove_initializer(replaced_matmul.input[1])
+                elif in_sn != replaced_matmul:
+                    out_sg.node.add().CopyFrom(in_sn)
+
+            scan = nf.make_node('Scan', new_inputs,
+                                {'body':out_sg,
+                                  'scan_input_directions':scan_input_directions,
+                                  'scan_output_directions':scan_output_directions,
+                                  'num_scan_inputs':num_scan_inputs},
+                                output_names=list(in_n.output))
+        else:
+            out_n = out_mp.graph.node.add()
+            out_n.CopyFrom(in_n)
+
+    onnx.save(out_mp, output_model)
+
+def parse_arguments():
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--mode', help='The modification mode', choices=['to_scan', 'opt_inproj'])
+  parser.add_argument('--input', help='The input model file', default=None)
+  parser.add_argument('--output', help='The output model file', default=None)
+  return parser.parse_args()
```

Comment:
Is there something somewhere that describes what this script does, why it is needed, and when it should be used?

Reply:
The explanation is in docs/execution_providers/Nuphar-ExecutionProvider.md. Will add another README.md in this folder.

---
In reply to: [312419093](https://github.com/microsoft/onnxruntime/pull/881#discussion_r312419093) [](ancestors = 312419093)